lib/queuePopulator/QueuePopulator.js

```diff
@@ -524,7 +524,21 @@ class QueuePopulator {
     _processLogEntries(params, done) {
         return async.map(
             this.logReaders,
-            (logReader, cb) => logReader.processLogEntries(params, cb),
+            (logReader, readerDone) => {
+                const batchCb = (err, hasMoreLog) => {
+                    if (err) {
+                        return readerDone(err);
+                    }
+                    if (!this.qpConfig.exhaustLogSource || this.logReadersUpdate || !hasMoreLog) {
+                        return readerDone();
+                    }
+                    this.log.debug('skipping wait, more logs are available', {
+                        method: 'QueuePopulator._processLogEntries',
+                    });
+                    return logReader.processLogEntries(params, batchCb);
```

Comment:
Is the callback recursion done on purpose here? We call batchCb from within the variable. If this goes for too long, without proper limit, we might have a stack overflow

Reply:
Good point, although this implementation is how it's done in S3C and it worked for the most part, it's no reason to keep it this way.